frontend/src/stores/media/single-result.ts

```diff
@@ -66,99 +84,136 @@ export const useSingleResultStore = defineStore("single-result", {
      *
-     * @param type - media type of the item to fetch.
-     * @param id - string id of the item to fetch.
+     * Fetch the related media if necessary.
      */
-    async fetchMediaItem(type: SupportedMediaType, id: string) {
-      try {
-        this._updateFetchState("start")
-        const accessToken = this.$nuxt.$openverseApiToken
-        const service = initServices[type](accessToken)
-        const item = this._addProviderName(await service.getMediaDetail(id))
-
-        // Fake ~50% of results as mature. This leaves actual mature results unchanged.
-        const featureFlagStore = useFeatureFlagStore()
-        if (featureFlagStore.isOn("fake_sensitive")) {
-          markFakeSensitive(item)
-        }
-
-        this.mediaItem = item
-        this.mediaType = type
+    async fetch<T extends SupportedMediaType>(
+      type: T,
+      id: string,
+      options?: { fetchRelated: boolean }
+    ) {
+      let item: DetailFromMediaType<T>
 
-        this._updateFetchState("end")
-      } catch (error: unknown) {
-        this.mediaItem = null
-        this.mediaType = type
-        if (axios.isAxiosError(error) && error.response?.status === 404) {
-          throw new Error(`Media of type ${type} with id ${id} not found`)
-        } else {
-          this.handleMediaError(error)
+      if (this.mediaId === id && isMediaDetail<T>(this.mediaItem, type)) {
+        item = this.mediaItem
+      } else {
+        try {
+          item = (await this.fetchMediaItem<T>(
+            type,
+            id
+          )) as DetailFromMediaType<T>
+        } catch (error) {
+          this.reset()
+          const statusCode =
+            axios.isAxiosError(error) && error.response?.status
+              ? error.response.status
+              : 404
+          const message = `Could not fetch ${type} item with id ${id}`
+          this._updateFetchState("end", {
+            statusCode,
+            message,
+          })
+          this.$nuxt.$sentry.captureException(error)
+          return null
         }
       }
+
+      const shouldFetchRelated = options?.fetchRelated ?? true
+      if (shouldFetchRelated) {
+        await this.fetchRelatedMedia(type, id)
+      }
+      return item
     },
 
     /**
-     * Throws a new error with a new error message.
+     * Fetch the media item from the API.
      */
-    handleMediaError(error: unknown) {
-      let errorMessage
-      if (axios.isAxiosError(error)) {
-        errorMessage =
-          error.response?.status === 500
-            ? "There was a problem with our servers"
-            : `Request failed with status ${
-                error.response?.status ?? "unknown"
-              }`
-      } else {
-        errorMessage =
-          error instanceof Error ? error.message : "Oops! Something went wrong"
+    async fetchMediaItem<T extends SupportedMediaType>(
+      type: SupportedMediaType,
+      id: string
+    ): Promise<DetailFromMediaType<typeof type>> {
```

Comment:
The generic `T` type can be used to remove the need to cast at call sites:

```suggestion
    async fetchMediaItem<MediaType extends SupportedMediaType>(
      type: MediaType,
      id: string
    ): Promise<DetailFromMediaType<MediaType>> {
```

Changing it from `T` to `MediaType` helps document what the type is.

The reason the cast was necessary at call sites is because `typeof type` resolved to broad `SupportedMediaType` rather than the specific type passed to `T` because `type` was typed as `SupportedMeidaType` rather than the generic type

Reply:
Nice! Thank you!